pkg/apis/node/types.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// RuntimeClass defines a class of container runtime supported in the cluster.
+// The RuntimeClass is used to determines which container runtime is used to run
+// all containers in a pod. RuntimeClasses are (currently) manually defined by a
+// user or cluster provisioner, and referenced in the PodSpec. The Kubelet is
+// responsible for resolving the RuntimeClassName reference before running the
+// pod.  For more details, see
+// https://git.k8s.io/enhancements/keps/sig-node/runtime-class.md
+type RuntimeClass struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ObjectMeta
+
+	// Specification of the RuntimeClass
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	Spec RuntimeClassSpec
+}
+
+// RuntimeClassSpec is a specification of a RuntimeClass. It contains parameters
+// that are required to describe the RuntimeClass to the Container Runtime
+// Interface (CRI) implementation, as well as any other components that need to
+// understand how the pod will be run. The RuntimeClassSpec should be treated as
+// immutable.
+type RuntimeClassSpec struct {
+	// RuntimeHandler specifies the underlying runtime and configuration that the
+	// CRI implementation will use to handle pods of this class. The possible
+	// values are specific to the node configuration & CRI implementation.  It is
+	// assumed that all handlers are available on every node, and handlers of the
+	// same name are equivalent on every node.
+	// For example, a handler called "runc" might specify that the runc OCI
+	// runtime (using native Linux containers) should be used to run the
+	// containers in a pod.
+	// If this is not specified (or empty), a default will be chosen by the
```

Comment:
Let's talk about defaults.  This is different than storage and as such raises alarms.

The way storage works is that there's an admission controller that looks at all StorgaeClass resources for a label, and the one that is labelled is the default.  If a volume doesn't have a class defined, the admission will set the default.  This is nice because a) the admission could be replaced by something more policy-ish and b) it's changeable fairly trivially.

It has the side-effect that volumes which already exist do not get their class set, which is actually correct.

Now, pods that are already running *do* have a class, even if it is hidden today.  Until this struct carries parameters, that is, at which point you don't know if it matches the class in reality.

It would be more consistent to go the storage route.  Define admission and a label, decorate incoming pods.  Make this field required.

Why should I not require that?

Reply:
> The way storage works is that there's an admission controller that looks at all StorgaeClass resources for a label, and the one that is labelled is the default.

This means that you can't have different defaults per namespace, and gets weird if you label multiple classes the default, or want to roll out a new default.

> If a volume doesn't have a class defined, the admission will set the default. This is nice because a) the admission could be replaced by something more policy-ish and b) it's changeable fairly trivially.

I am expecting that the default RuntimeClass will be a policy decision (e.g. via PodSecurityPolicy: https://github.com/kubernetes/kubernetes/pull/73795).

> It has the side-effect that volumes which already exist do not get their class set, which is actually correct.

Right, that's what this is addressing. If a pod does not have a runtimeclass (i.e. `pod.spec.runtimeClassName == nil`), then the CRI implementation has the final say. From the CRI perspective, that is equivalent to a pod using a RuntimeClass with a `nil` runtimeHandler. We could make runtimeHandler a required field in the RuntimeClass, but then there is no way for a RuntimeClass to target the backwards-compatible "legacy" handler. Being able to target that node-level default may be useful once we roll out other control-plane level features like PodOverhead & scheduling.

Does that make sense? If not, maybe we could grab a few minutes face-to-face.

Reply:
From IRL discussion, containerd (for example) has an unnamed default.  I'd like to request that containerd treat "" and "default" (or something) equivalently, and make this field required.

That has a side-effect that alpha has it as optional.  I think it should be OK to require this in beta as long as you satisfy the round-trip, which you'll need an annotation for.  E.g.

alpha { handler: "" } => beta { annotations: { "alpha-handler-was-empty": "true" }, handler: "default" }


@liggitt to see if he pukes on that idea.  I would never suggest it for anything but alpha compat.

Reply:
Conclusion from offline discussion: make this field required, reserve a name for the behavior when no RuntimeClass is used.

Open questions:
- should this resolution of the reserved name be done by the kubelet (i.e. pass a nil value to the CRI), or should the resolution happen in the CRI itleself?
- what should the reserved value be? `default` or `legacy`?
- Is it ok to move this to a required field going from alpha to beta? What about going from CRD to bulit-in? @lavalamp @liggitt 

@kubernetes/sig-node-api-reviews

Reply:
missing the offline context, but as far as mechanics are concerned, an optional field in alpha can easily be required in beta, as long as the alpha version has a default (defaults are per-version).

that looks like:
* alpha defaults
* beta doesn't default
* validation requires a value

Reply:
Since the resources need to be recreated anyway for the CRD -> internal transition, is defaulting still needed? I was thinking of just making the field required in the built-in v1alpha1. Sure it's breaking, but so is moving to an internal API....

> reserve a name for the behavior when no RuntimeClass is used

Thinking about this a bit more, I think we should scrap this. Handler should be explicit.